pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/QueryDispatcher.java

```diff
@@ -139,22 +140,24 @@ int submit(long requestId, DispatchableSubPlan dispatchableSubPlan, long timeout
         for (Map.Entry<QueryServerInstance, List<Integer>> queryServerEntry
             : dispatchableSubPlan.getQueryStageList().get(stageId).getServerInstanceToWorkerIdMap().entrySet()) {
           QueryServerInstance queryServerInstance = queryServerEntry.getKey();
+          Worker.QueryRequest.Builder queryRequestBuilder = Worker.QueryRequest.newBuilder();
+          String host = queryServerInstance.getHostname();
+          int servicePort = queryServerInstance.getQueryServicePort();
+          int mailboxPort = queryServerInstance.getQueryMailboxPort();
           for (int workerId : queryServerEntry.getValue()) {
-            String host = queryServerInstance.getHostname();
-            int servicePort = queryServerInstance.getQueryServicePort();
-            int mailboxPort = queryServerInstance.getQueryMailboxPort();
             VirtualServerAddress virtualServerAddress = new VirtualServerAddress(host, mailboxPort, workerId);
-            DispatchClient client = getOrCreateDispatchClient(host, servicePort);
             dispatchCalls++;
-            int finalStageId = stageId;
-            _executorService.submit(() -> client.submit(Worker.QueryRequest.newBuilder().setStagePlan(
-                        QueryPlanSerDeUtils.serialize(
-                            constructDistributedStagePlan(dispatchableSubPlan, finalStageId, virtualServerAddress)))
-                    .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_ID, String.valueOf(requestId))
-                    .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS, String.valueOf(timeoutMs))
-                    .putAllMetadata(queryOptions).build(), finalStageId, queryServerInstance, deadline,
-                dispatchCallbacks::offer));
+            queryRequestBuilder.addStagePlan(
+                QueryPlanSerDeUtils.serialize(dispatchableSubPlan, stageId, virtualServerAddress));
```

Comment:
This would serialize the entire sub plan for each worker and stageId combination. Big use-cases would usually have 256 partitions (i.e. workers) and ~10 or so stages.

So this means we'll call this ~2000 times. For low qps use-cases this should be fine but relatively higher use-cases might start getting bottlenecked.

@walterddr : I remember we had discussed this a couple of months ago. Was wondering if we are planning to fix this anytime soon. cc: @xiangfu0

Reply:
This is good point, I added one more commit to this PR, you can check the diff: https://github.com/apache/pinot/pull/10791/commits/88ce2598618dfc69f7db96302e6983303962cccc

Reply:
The basic idea is to carry all the workIds within one stagePlan then expand it at the worker side to generate one DistributedStagePlan per worker.

Reply:
let's consider separate address this issue into a different PR.
this one is already quite large and hard to review. 

i guess the problem i have with this is it adds unnecessary info to the proto

Reply:
actually we decided to include this in the PR, b/c both changes require proto modification so it is best to avoid multiple changes to wire protocol. updated the PR by force pushing